File: apps/ios/Pulse/Features/VoiceRooms/VoicePcmChunker.swift
import Foundation

// ─────────────────────────────────────────────────────────────
// Pulse — W5-f voice room PCM chunker (PURE, no AVFoundation).
//
// The capture contract (spec §1.1 VR-4, web voice-room-sheet.tsx):
//   · 16 kHz mono Int16 input
//   · full chunks every 250 ms = 4000 samples, base64(Int16LE bytes)
//   · seq starts at 1 and increments per emitted chunk
//   · a partial block (< 4000 samples) is flushed PROPORTIONALLY on
//     release — the emitted chunk simply carries the remaining
//     samples (its duration is the proportional fraction of 250 ms),
//     never padded, never dropped
//   · reset() returns the machine to the cold state (seq = 1)
//
// WEB DEFECT FIX #1: the web sheet armed `blockRef.buf` but never fed
// it from the capture callback, so NO voice:chunk ever left the
// client. This chunker is the single feed path for captured samples
// and its chunk/seq/partial/reset semantics are proof-tested in
// VoiceRoomMachineTests (the fix is proven by test, per spec VR-4).
// ─────────────────────────────────────────────────────────────

/// One emitted chunk — the exact C→S `voice:chunk` fields.
public struct VoicePcmChunkOut: Equatable, Sendable {
    public let seq: Int
    public let base64: String

    public init(seq: Int, base64: String) {
        self.seq = seq
        self.base64 = base64
    }
}

/// Pure block machine — feed Int16 samples, receive base64 chunks.
/// A value type: the audio engine owns one instance and mutates it on
/// its capture queue; unit tests drive it directly.
public struct VoicePcmChunker: Sendable {
    /// Samples per full chunk — 16 kHz × 250 ms.
    public let chunkSamples: Int
    private var buffer: [Int16] = []
    private var nextSeq: Int

    /// Cold state: empty buffer, first chunk will carry seq 1.
    public init(chunkSamples: Int = 4000) {
        self.chunkSamples = chunkSamples
        self.nextSeq = 1
    }

    /// Samples currently held (not yet part of an emitted chunk).
    public var pendingCount: Int { buffer.count }

    /// Next seq the machine would emit (1 before the first chunk).
    public var upcomingSeq: Int { nextSeq }

    /// True when at least one full chunk is waiting to be drained.
    public var hasFullChunk: Bool { buffer.count >= chunkSamples }

    /// Feeds samples and drains every COMPLETED block. Order is
    /// guaranteed: chunks come out in seq order, one per 4000 samples.
    public mutating func feed(_ samples: [Int16]) -> [VoicePcmChunkOut] {
        guard !samples.isEmpty else { return [] }
        buffer.append(contentsOf: samples)
        var out: [VoicePcmChunkOut] = []
        while buffer.count >= chunkSamples {
            let block = Array(buffer.prefix(chunkSamples))
            buffer.removeFirst(chunkSamples)
            out.append(Self.encode(seq: nextSeq, samples: block))
            nextSeq += 1
        }
        return out
    }

    /// Proportional partial flush on PTT release — emits whatever is
    /// left (< one full block) as the FINAL chunk of the burst. Nil
    /// when nothing is pending (the burst ended on a block boundary).
    public mutating func flushPartial() -> VoicePcmChunkOut? {
        guard !buffer.isEmpty else { return nil }
        let chunk = Self.encode(seq: nextSeq, samples: buffer)
        buffer.removeAll()
        nextSeq += 1
        return chunk
    }

    /// Teardown (VR-9): the next burst starts from seq 1 again.
    public mutating func reset() {
        buffer.removeAll()
        nextSeq = 1
    }

    /// Int16LE little-endian bytes → base64 (the wire `data` field).
    private static func encode(seq: Int, samples: [Int16]) -> VoicePcmChunkOut {
        var bytes = Data(capacity: samples.count * 2)
        for sample in samples {
            var le = sample.littleEndian
            withUnsafeBytes(of: &le) { bytes.append(contentsOf: $0) }
        }
        return VoicePcmChunkOut(seq: seq, base64: bytes.base64EncodedString())
    }
}

File: apps/android/feature-voice/src/main/kotlin/app/pulse/feature/voice/engine/VoicePcmChunker.kt
package app.pulse.feature.voice.engine

/**
 * The 250ms transmit block builder (spec §1.1 VR-4 / WEB DEFECT FIX #1).
 *
 * The web's CRITICAL defect was a transmit path that never armed its block
 * buffer — no voice:chunk ever left the browser. This chunker is the native
 * equivalent and its unit tests are the PROOF that chunks actually flow:
 * 16kHz mono Int16 in, 4000-sample base64-bound blocks out, seq starting at 1,
 * a proportional partial flush on release, and a hard reset for new sessions.
 *
 * Seq semantics: seq is CONTINUOUS for the whole room session (join → leave),
 * NOT per transmission — restarting at 1 on every push would trip every
 * receiver's `seq <= lastSeq` dedupe and blackhole the audio (the exact web
 * rejoin defect the roster-reset fix answers on the receive side).
 */
class VoicePcmChunker(private val blockSize: Int = VoicePcm.BLOCK_SAMPLES) {

    /** One ready-to-emit block: [seq] starts at 1; [samples] ≤ [blockSize]. */
    class Chunk(val seq: Long, val samples: ShortArray)

    private var nextSeq = 1L
    private var window = ShortArray(blockSize * 2)
    private var windowSize = 0

    /** Total samples accepted since the last [reset] — test/diagnostic exposure. */
    var consumedSamples: Long = 0L
        private set

    /**
     * Feeds captured samples; returns every FULL block that became available.
     * Remainders stay buffered until they fill or [flush] is called.
     */
    fun consume(samples: ShortArray): List<Chunk> {
        val chunks = mutableListOf<Chunk>()
        var offset = 0
        while (offset < samples.size) {
            val space = blockSize - windowSize
            val take = minOf(space, samples.size - offset)
            if (windowSize + take > window.size) {
                window = window.copyOf(maxOf(blockSize, (windowSize + take) * 2))
            }
            System.arraycopy(samples, offset, window, windowSize, take)
            windowSize += take
            offset += take
            if (windowSize == blockSize) {
                chunks += Chunk(nextSeq++, window.copyOf(blockSize))
                windowSize = 0
            }
        }
        consumedSamples += samples.size
        return chunks
    }

    /**
     * Release-time flush: any PARTIAL window emits as one proportionally
     * smaller block (never padded with silence), then the window resets and
     * the sequence keeps counting. Null when there is nothing partial.
     */
    fun flush(): Chunk? {
        if (windowSize == 0) return null
        val chunk = Chunk(nextSeq++, window.copyOf(windowSize))
        windowSize = 0
        return chunk
    }

    /** New session (join): clears the window and restarts the sequence at 1 (VR-9). */
    fun reset() {
        nextSeq = 1
        windowSize = 0
        consumedSamples = 0
    }
}
